### PR TITLE
Revert "net: tcp: Fix possible deadlock in tcp_conn_unref()"

### PR DESCRIPTION
**Pull request by neja20**
Reverts zephyrproject-rtos/zephyr#42526

As I am testing this change, it seems that the net_ctx is getting leaked. So it seems it may not be the rightway to unlock the tcp_lock.

I would like to consider the option of manipulating `ctx->cond.lock` lock. For example get rid of `(void)k_mutex_lock(ctx->cond.lock, K_FOREVER);` in `static void zsock_received_cb()`. How do u feel about that? Because it seems to work better than the other way

### CODE REVIEW

**Comment by nashif** [comment]
you need a signoff in the commit message.

**Comment by rlubos** [comment]
> I would like to consider the option of manipulating ctx->cond.lock lock. For example get rid of (void)k_mutex_lock(ctx->cond.lock, K_FOREVER); in static void zsock_received_cb().

I don't think we should remove that mutex protection - it was added for a reason, removing it could lead to other issues with recv queue handling in the socket/net_context.

I'm thinking about the `tcp_conn_unref()` function - it seems to lock the mutex first, and only after that it decrements the ref count with  `atomic_dec()` (which should be safe w/o a mutex protection) and check if it actually needs to release the tcp context (and this part should be protected by the mutex in my understanding).

Maybe a good solution would be to move this check before the mutex lock - only one thread should attempt to lock the tcp mutex then, therefore preventing the deadlock. It's implemented in a similar way in net_context (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/net_context.c#L361).

**Comment by rlubos** [comment]
CC @jukkar

**Comment by neja20** [comment]
Thanks @rlubos , that sounds reasonable. I will take a look at that

**Comment by neja20** [comment]
hi,

@rlubos's suggestions led me to implement the `tcp_conn_unref()` similary to https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/net_context.c#L361, which definitely helped and revealed the hanging lock in the sockets.

There is a condition variable `ctx->cond.recv` which I assume is being used as a replacement to Events, but the k_condvar_wait() reaquires the lock after the wait is finished and the lock was never released which was the other reason why could tcp stack's `rx_q` thread be locked infinitely, hence the `added k_mutex_unlock()` call.

File: subsys/net/ip/tcp.c [hunk 1]
```diff
@@ -361,8 +361,12 @@ static int tcp_conn_unref_debug(struct tcp *conn, const char *caller, int line)
 static int tcp_conn_unref(struct tcp *conn)
 #endif
 {
-	int ref_count = atomic_get(&conn->ref_count);
 	struct net_pkt *pkt;
+	int ref_count = atomic_dec(&conn->ref_count) - 1;
```
Comment (rlubos) [inline subsys/net/ip/tcp.c:365]: And move the decrement and check just before the mutex lock, so that we don't check other logic.

File: subsys/net/lib/sockets/sockets.c [hunk 2]
```diff
@@ -980,8 +980,13 @@ int zsock_wait_data(struct net_context *ctx, k_timeout_t *timeout)
 
 	if (k_fifo_is_empty(&ctx->recv_q)) {
 		/* Wait for the data to arrive but without holding a lock */
-		return k_condvar_wait(&ctx->cond.recv, ctx->cond.lock,
+		int ret = k_condvar_wait(&ctx->cond.recv, ctx->cond.lock,
 				      *timeout);
+		/* Release the lock because the condition variable is
+		 * just used for signaling purpose.
+		 */
+		k_mutex_unlock(ctx->cond.lock);
```
Comment (rlubos) [inline subsys/net/lib/sockets/sockets.c:988]: I don't think it's right - one of the purpose of `k_condvar_wait()` is to restore the lock once it quits - the socket lock will be released in the corresponding `recv()`/`recvfrom()` call.

Comment (neja20) [reply subsys/net/lib/sockets/sockets.c:988]: You are right that the lock is aquiered when calling `recv()`/`recvfrom()` but when calling `read()` it enters the `k_condvar_wait()` without previously owning the lock so that's where the uneven number of lock/unlock is happening.

Is this behavior corret? Should I only call `recv()` and not `read()` from my app?

Comment (rlubos) [reply subsys/net/lib/sockets/sockets.c:988]: Ok, that's not right then, `read()` should also acquire the mutex. It seems that this should be addressed in the `fdtable.c` which provides generic `read()` implementation (I think the mutex should be acquired before vtable call, just like in case other calls).

**Comment by carlescufi** [comment]
@neja20 you don't need to close the PR, just force-push to your branch instead.

**Comment by neja20** [comment]
@carlescufi I am sorry, I might be using git the wrong way. I will find a way to do this properly next time.

File: subsys/net/lib/sockets/sockets.c (hunk 2 above)
Comment (neja20) [reply subsys/net/lib/sockets/sockets.c]: These changes in PR works well for me. I dont't know if I should add the locking to other functions working with vtable like `write()`, `close()`?

Comment (rlubos) [reply subsys/net/lib/sockets/sockets.c]: I think it'd make sense, both for consistency and additional protection.

**Comment by carlescufi** [comment]
> @carlescufi I am sorry, I might be using git the wrong way. I will find a way to do this properly next time.

The usual workflow is: 

```
<make changes>
$ git add
$ git commit --amend
$ git push --force
```

File: subsys/net/ip/tcp.c [hunk 3]
```diff
@@ -377,15 +377,13 @@ static int tcp_conn_unref(struct tcp *conn)
 	}
 #endif /* CONFIG_NET_TEST_PROTOCOL */
 
-	k_mutex_lock(&tcp_lock, K_FOREVER);
-
 	ref_count = atomic_dec(&conn->ref_count) - 1;
-	if (ref_count) {
-		tp_out(net_context_get_family(conn->context), conn->iface,
-		       "TP_TRACE", "event", "CONN_DELETE");
```
Comment (rlubos) [inline subsys/net/ip/tcp.c:385]: You shouldn't delete these lines.

**Review by rlubos — COMMENTED** [review]
One more comment.

**Review by rlubos — APPROVED** [review]
Thanks, looking good.

File: subsys/net/ip/tcp.c [hunk 4]
```diff
@@ -377,15 +377,15 @@ static int tcp_conn_unref(struct tcp *conn)
...
 
-	k_mutex_lock(&tcp_lock, K_FOREVER);
-
 	ref_count = atomic_dec(&conn->ref_count) - 1;
-	if (ref_count) {
+	if (ref_count != 0) {
 		tp_out(net_context_get_family(conn->context), conn->iface,
-		       "TP_TRACE", "event", "CONN_DELETE");
-		goto unlock;
+			"TP_TRACE", "event", "CONN_DELETE");
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:383]: Unnecessary change in this line

File: lib/os/fdtable.c [hunk 5]
```diff
@@ -224,21 +224,37 @@ int z_alloc_fd(void *obj, const struct fd_op_vtable *vtable)
```
Comment (jukkar) [inline lib/os/fdtable.c:224]: Typo in commit message of `lib/os: fdtable: add locking to posix api`
s/sokcets/sockets/

**Review by jukkar — COMMENTED** [review]
Looks reasonable, couple of minor nits that could be fixed.

File: subsys/net/ip/tcp.c (hunk 4 above)
Comment (neja20) [reply subsys/net/ip/tcp.c]: Fixed

File: lib/os/fdtable.c (hunk 5 above)
Comment (neja20) [reply lib/os/fdtable.c:224]: Fixed

**Review by jukkar — APPROVED** [review]
